Allow docker pause and unpause to accept multiple containers

aanm opened the pull request:
Applied multi parameters to pause and unpause.
Created a new test file dedicated for pause commands.
Created a new utility function to get a slice of paused containers.

Fixes #10457

Signed-off-by: André Martins martins@noironetworks.com

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,95 @@
+	image := strings.Split(out, "\n")[0]
+	dockerCmd(t, "run", "-d", "--name", name, image, "sleep", "2")
+	dockerCmd(t, "pause", name)
+	dockerCmd(t, "unpause", name)
```
Can you use the getPausedContainers() func to actually get the list of paused containers?  I know you check the event log later on but I'd like to actually check the current state (between the pause and unpause) to make sure things are really paused.

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,95 @@
+	dockerCmd(t, "pause", name)
+	dockerCmd(t, "unpause", name)
+
+	defer deleteAllContainers()
```
Can move this defer to the top then if things die during the pause/unpause we'll still clean-up?

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,95 @@
+	dockerCmd(t, append([]string{"pause"}, containers...)...)
+	dockerCmd(t, append([]string{"unpause"}, containers...)...)
+
+	defer deleteAllContainers()
```
move to the top

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,95 @@
+		dockerCmd(t, "run", "-d", "--name", name, image, "sleep", "2")
+	}
+	dockerCmd(t, append([]string{"pause"}, containers...)...)
+	dockerCmd(t, append([]string{"unpause"}, containers...)...)
```
same as above, use the getPausedContainers() to make sure they're really paused.

crosbymichael commented:
@aanm i edited the title and added the fixes line to the description in the PR

crosbymichael commented:
It looks like the CI is mad because you need to go fmt the files where you made changes to fix the formatting.  

Thanks for working on this so quickly!

aanm committed:
Fixes #10457-Pause-and-unpause-accept-multi-containers

Applied multi parameters to pause and unpause.
Created a new test file dedicated for pause commands.
Created a new utility function to get a slice of paused containers.
Updated documentation

Signed-off-by: André Martins <martins@noironetworks.com>

SvenDowideit commented:
Can you please update the documentation for pause and unpause (both cli.md and the docs/man pages)

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+	"time"
+)
+
+func TestEventsPause(t *testing.T) {
```
Can you rename this to TestPause since its no longer part of the docker_cli_events_test.go file ?

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+	logDone("events - pause/unpause is logged")
+}
+
+func TestEventsPauseMultipleContainers(t *testing.T) {
```
here too - remove the word "Events"

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+		t.Fatalf("error thrown while waiting for container: %s, %v", waitOut, err)
+	}
+
+	logDone("events - pause/unpause is logged")
```
s/events/pause/

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+		}
+	}
+
+	logDone("events - multi pause/unpause is logged")
```
s/events/pause/

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+	name := "testeventpause"
+	out, _, _ := dockerCmd(t, "images", "-q")
+	image := strings.Split(out, "\n")[0]
+	dockerCmd(t, "run", "-d", "--name", name, image, "sleep", "2")
```
I know this is how it was before your edits, but its interesting that the return values aren't checked.  I'm pretty sure things will fail anyway later on, but @jfrazelle is it normal practice to check stuff like this in tests?

duglin commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	for _, name := range containers {
+		waitCmd := exec.Command(dockerBinary, "wait", name)
```
any reason you didn't use: `docker wait name1 name2` ?

aanm replied:
No reason at all, only because

jfrazelle commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -30,41 +30,6 @@ func TestEventsUntag(t *testing.T) {
 	logDone("events - untags are logged")
 }
 
-func TestEventsPause(t *testing.T) {
```
why is this test being removed

duglin replied:
I believe it was moved into a new "pause" test file.

jfrazelle replied:
oh my bad it was renamed too

jfrazelle commented on integration-cli/docker_cli_pause_test.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func TestPauseMultipleContainers(t *testing.T) {
+	defer deleteAllContainers()
```
we need to also defer unpauseAllContainers() or else we cant send a sig

jfrazelle replied:
same goes for the above test

duglin replied:
wow - deleteAllContainers() won't kill paused ones?  I didn't know that.

jfrazelle replied:
well so paused containers we cant send a sig to, its why we didnt have
pause tests for so long

On Fri, Feb 6, 2015 at 3:55 PM, Doug Davis <notifications@github.com> wrote:

> In integration-cli/docker_cli_pause_test.go
> <https://github.com/docker/docker/pull/10468#discussion_r24281597>:
>
> > +		t.Fatalf("event should be pause, not %#v", pauseEvent)
> > +	}
> > +	if unpauseEvent[len(unpauseEvent)-1] != "unpause" {
> > +		t.Fatalf("event should be unpause, not %#v", unpauseEvent)
> > +	}
> > +
> > +	waitCmd := exec.Command(dockerBinary, "wait", name)
> > +	if waitOut, _, err := runCommandWithOutput(waitCmd); err != nil {
> > +		t.Fatalf("error thrown while waiting for container: %s, %v", waitOut, err)
> > +	}
> > +
> > +	logDone("pause - pause/unpause is logged")
> > +}
> > +
> > +func TestPauseMultipleContainers(t *testing.T) {
> > +	defer deleteAllContainers()
>
> wow - deleteAllContainers() won't kill paused ones? I didn't know that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10468/files#r24281597>.
>

jfrazelle commented:
LGTM

jfrazelle commented:
ping @duglin @crosbymichael ;)

duglin commented:
Yup LGTM

tiborvass commented:
LGTM Ping @fredlf @SvenDowideit

SvenDowideit commented:
Docs LGTM - @fredlf @jamtur01

jamtur01 commented on docs/sources/reference/commandline/cli.md:
```diff
@@ -1396,22 +1396,6 @@ just a specific mapping:
-See the
-[cgroups freezer documentation](https://www.kernel.org/doc/Documentation/cgroups/freezer-subsystem.txt)
-for further details.
-
```
Any reason this detail was removed?

aanm replied:
I don't know it was there. I remove it because there was another "pause" on lines 1362-1375

SvenDowideit replied:
@jamtur01 its an out of alpha-order duplicate.

SvenDowideit commented:
@fredlf @jamtur01

fredlf commented:
LGTM
